In [53]:
import pandas as pd

df = pd.read_csv("SampleSuperstore.csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)

df.head()

Dataset loaded successfully!
Shape: (9994, 13)


,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164


In [54]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes)

print("\nMissing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:", df.duplicated().sum())

Number of rows: 9994
Number of columns: 13

Column Names:
['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount', 'Profit']

Data Types:
Ship Mode        object
Segment          object
Country          object
City             object
State            object
Postal Code       int64
Region           object
Category         object
Sub-Category     object
Sales           float64
Quantity          int64
Discount        float64
Profit          float64
dtype: object

Missing Values:
Ship Mode       0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Category        0
Sub-Category    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

Duplicate Rows: 17


In [55]:
# Data Quality Audit

audit_results = []

# 1. Missing values
for column in df.columns:
    missing_count = df[column].isnull().sum()
    audit_results.append({
        "Check": "Missing Values",
        "Column": column,
        "Issue Count": missing_count
    })

# 2. Duplicate rows
duplicate_count = df.duplicated().sum()
audit_results.append({
    "Check": "Duplicate Rows",
    "Column": "All Columns",
    "Issue Count": duplicate_count
})

# 3. Invalid Quantity
invalid_quantity = (df["Quantity"] <= 0).sum()
audit_results.append({
    "Check": "Invalid Quantity",
    "Column": "Quantity",
    "Issue Count": invalid_quantity
})

# 4. Invalid Sales
invalid_sales = (df["Sales"] < 0).sum()
audit_results.append({
    "Check": "Invalid Sales",
    "Column": "Sales",
    "Issue Count": invalid_sales
})

# 5. Invalid Discount
invalid_discount = ((df["Discount"] < 0) | (df["Discount"] > 1)).sum()
audit_results.append({
    "Check": "Invalid Discount",
    "Column": "Discount",
    "Issue Count": invalid_discount
})

# 6. Invalid Postal Code
invalid_postal = (df["Postal Code"] <= 0).sum()
audit_results.append({
    "Check": "Invalid Postal Code",
    "Column": "Postal Code",
    "Issue Count": invalid_postal
})

# Convert results to DataFrame
audit_df = pd.DataFrame(audit_results)

audit_df

,Check,Column,Issue Count
0,Missing Values,Ship Mode,0
1,Missing Values,Segment,0
2,Missing Values,Country,0
3,Missing Values,City,0
4,Missing Values,State,0
5,Missing Values,Postal Code,0
6,Missing Values,Region,0
7,Missing Values,Category,0
8,Missing Values,Sub-Category,0
9,Missing Values,Sales,0


In [56]:
# Consistency Checks

print("Ship Mode values:")
print(df["Ship Mode"].unique())

print("\nSegment values:")
print(df["Segment"].unique())

print("\nCountry values:")
print(df["Country"].unique())

print("\nRegion values:")
print(df["Region"].unique())

print("\nCategory values:")
print(df["Category"].unique())

print("\nSub-Category values:")
print(df["Sub-Category"].unique())

Ship Mode values:
['Second Class' 'Standard Class' 'First Class' 'Same Day']

Segment values:
['Consumer' 'Corporate' 'Home Office']

Country values:
['United States']

Region values:
['South' 'West' 'Central' 'East']

Category values:
['Furniture' 'Office Supplies' 'Technology']

Sub-Category values:
['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']


In [57]:
# Create Issue Log

issue_log = pd.DataFrame({
    "Issue Type": [
        "Duplicate Rows"
    ],
    "Column": [
        "All Columns"
    ],
    "Issue Count": [
        17
    ],
    "Severity": [
        "Medium"
    ],
    "Action Taken": [
        "Duplicate rows removed"
    ]
})

issue_log

,Issue Type,Column,Issue Count,Severity,Action Taken
0,Duplicate Rows,All Columns,17,Medium,Duplicate rows removed


In [58]:
issue_log.to_csv("Issue_Log.csv", index=False)

print("Issue Log saved successfully!")

Issue Log saved successfully!


In [59]:
# Remove duplicate rows

cleaned_df = df.drop_duplicates().copy()

print("Original rows:", len(df))
print("Rows after removing duplicates:", len(cleaned_df))
print("Rows removed:", len(df) - len(cleaned_df))

Original rows: 9994
Rows after removing duplicates: 9977
Rows removed: 17


In [60]:
# Final Quality Check

print("Final Dataset Shape:", cleaned_df.shape)

print("\nMissing Values:")
print(cleaned_df.isnull().sum())

print("\nDuplicate Rows:", cleaned_df.duplicated().sum())

print("\nInvalid Quantity:", (cleaned_df["Quantity"] <= 0).sum())

print("\nInvalid Sales:", (cleaned_df["Sales"] < 0).sum())

print("\nInvalid Discount:",
      ((cleaned_df["Discount"] < 0) | (cleaned_df["Discount"] > 1)).sum())

print("\nInvalid Postal Code:",
      (cleaned_df["Postal Code"] <= 0).sum())

Final Dataset Shape: (9977, 13)

Missing Values:
Ship Mode       0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Category        0
Sub-Category    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

Duplicate Rows: 0

Invalid Quantity: 0

Invalid Sales: 0

Invalid Discount: 0

Invalid Postal Code: 0


Data Quality Audit Report

Dataset Summary
- Original Records: 9,994
- Columns: 13
- Final Records: 9,977

Quality Checks
| Quality Check | Issues Found |
|---|---:|
| Missing Values | 0 |
| Duplicate Rows | 17 |
| Invalid Quantity | 0 |
| Invalid Sales | 0 |
| Invalid Discount | 0 |
| Invalid Postal Code | 0 |
| Consistency Issues | 0 |

Cleaning Performed
- 17 duplicate rows were identified and removed.
- No missing values were found.
- No invalid quantity, sales, discount, or postal code values were found.
- No categorical consistency issues were identified.

Final Result

After cleaning, the dataset contains 9,977 records with no remaining duplicate rows or identified validation issues.

In [61]:
cleaned_df.to_csv("Task_24.csv", index=False)

print("Task 24 cleaned sample saved successfully!")

Task 24 cleaned sample saved successfully!


Final Summary

The Data Quality Audit was completed on the Sample Superstore dataset.

- Original records: 9,994
- Duplicate records identified: 17
- Duplicate records removed: 17
- Final records: 9,977
- Missing values: 0
- Invalid Quantity values: 0
- Invalid Sales values: 0
- Invalid Discount values: 0
- Invalid Postal Code values: 0
- Consistency issues: 0

The final quality check confirmed that no duplicate records or other identified validation issues remained after cleaning.